scheduler/reconcile_util.go

```diff
@@ -229,21 +229,25 @@ func (a allocSet) filterByTainted(taintedNodes map[string]*structs.Node, serverS
 		// without max_client_disconnect
 		supportsDisconnectedClients := alloc.SupportsDisconnectedClients(serverSupportsDisconnectedClients)
 
-		reconnected := false
+		reconnect := false
 		expired := false
 
-		// Only compute reconnected for unknown, running, and failed since they need to go through the reconnect logic.
+		// Only compute reconnect for unknown, running, and failed since they
+		// need to go through the reconnect logic.
 		if supportsDisconnectedClients &&
 			(alloc.ClientStatus == structs.AllocClientStatusUnknown ||
 				alloc.ClientStatus == structs.AllocClientStatusRunning ||
 				alloc.ClientStatus == structs.AllocClientStatusFailed) {
-			reconnected, expired = alloc.Reconnected()
+			reconnect = alloc.NeedsToReconnect()
+			if reconnect {
+				expired = alloc.Expired(time.Now())
```

Comment:
filterByTainted has `time.Now()` passed in to help with testing. 
```suggestion
				expired = alloc.Expired(now)
```